repository/src/interpret/read.rs

```diff
@@ -0,0 +1,259 @@
+        Ok(block_header)
+    } else {
+        Err(eyre!(IntegrityError {
+            msg: "`finalized` branch is not on a block".to_owned(),
+        }))
+    }
+}
+
+pub async fn read_last_finalization_proof(
+    raw: &RawRepository,
+) -> Result<LastFinalizationProof, Error> {
+    if let Ok(last_finalization_proof) = format::fp_from_semantic_commit(
+        raw.read_semantic_commit(raw.locate_branch(FP_BRANCH_NAME.into()).await.map_err(
+            |e| match e {
+                raw::Error::NotFound(_) => {
+                    eyre!(IntegrityError::new("cannot locate `fp` branch".to_string()))
+                }
+                _ => eyre!(e),
+            },
+        )?)
+        .await?,
+    ) {
+        Ok(last_finalization_proof)
+    } else {
+        Err(eyre!(IntegrityError {
+            msg: "`fp` branch is not on a finalization proof".to_owned(),
+        }))
+    }
+}
+
+pub async fn read_last_finalization_info(raw: &RawRepository) -> Result<FinalizationInfo, Error> {
+    let header = read_last_finalized_block_header(raw).await?;
+    let commit_hash = get_last_finalized_block_commit_hash(raw).await?;
+    let reserved_state = read_last_finalized_reserved_state(raw).await?;
+    let proof = read_last_finalization_proof(raw).await?.proof;
+    Ok(FinalizationInfo {
+        header,
+        commit_hash,
+        reserved_state,
+        proof,
+    })
+}
+
+pub async fn read_commits(
+    raw: &RawRepository,
+    ancestor: CommitHash,
+    descendant: CommitHash,
+) -> Result<Vec<(Commit, CommitHash)>, CommitError> {
+    let commits = raw.query_commit_path(ancestor, descendant).await?;
+    let commits = stream::iter(
+        commits
+            .iter()
+            .cloned()
+            .map(|c| async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }),
+    )
+    .buffered(256)
+    .collect::<Vec<_>>()
+    .await;
+    let commits = commits.into_iter().collect::<Result<Vec<_>, _>>()?;
+    let commits = commits
+        .into_iter()
+        .map(|(commit, hash)| {
+            from_semantic_commit(commit)
+                .map_err(|e| (e, hash))
+                .map(|x| (x, hash))
+        })
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|(e, c)| CommitError::Commit(e, c))?;
+    Ok(commits)
+}
+
+/// Reads the sequence of commits from the last finalized block to the given commit,
+/// and verifies them, and returns the CSV that all commits have been applied on.
+/// It does not accept a last finalization proof commit.
+///
+/// - Returns `Ok(Ok(commit))` if the branch (taking the given commit as a tip) is valid.
+/// - Returns `Ok(Err(error))` if the branch is invalid.
+/// - Returns `Err(error)` if there was an error reading the commits.
+pub async fn read_and_verify_commits_from_last_finalized_block(
+    raw: &RawRepository,
+    commit_hash: CommitHash,
+) -> Result<Result<CommitSequenceVerifier, Error>, Error> {
+    let lfi = read_last_finalization_info(raw).await?;
```

Comment:
Is `last_finalization_info` just too long for a variable name?

Reply:
I think so. Like `fp` or `dms`. I think it's worth having a doc for those abbreviations